File: ClassNotes/Services/AuthService.swift
import Foundation
import SwiftUI
import FirebaseAuth
import FirebaseFirestore

class AuthService: ObservableObject {
    @Published var isAuthenticated = false
    @Published var needsOnboarding = false
    @Published var currentUserId: String = ""
    @Published var currentUserName: String = ""

    private var verificationId: String?
    private lazy var db = Firestore.firestore()
    private var authStateListener: AuthStateDidChangeListenerHandle?
    private var fcmTokenObserver: NSObjectProtocol?

    init() {
        guard !AppMode.isDemo else { return }
        authStateListener = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            guard let self else { return }
            if let user {
                self.currentUserId = user.uid
                self.isAuthenticated = true
                self.fetchUserProfile(userId: user.uid)
                // Save FCM token for push notifications
                NotificationService.shared.updateFCMToken(for: user.uid)
            } else {
                self.currentUserId = ""
                self.currentUserName = ""
                self.isAuthenticated = false
            }
        }

        // Observe FCM token refreshes
        fcmTokenObserver = NotificationCenter.default.addObserver(
            forName: .fcmTokenRefreshed,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self, !self.currentUserId.isEmpty else { return }
            NotificationService.shared.updateFCMToken(for: self.currentUserId)
        }
    }

    deinit {
        if let listener = authStateListener {
            Auth.auth().removeStateDidChangeListener(listener)
        }
        if let observer = fcmTokenObserver {
            NotificationCenter.default.removeObserver(observer)
        }
    }

    // MARK: - Fetch User Profile

    private func fetchUserProfile(userId: String) {
        db.collection("users").document(userId).getDocument { [weak self] snapshot, error in
            guard let self else { return }
            DispatchQueue.main.async {
                if let data = snapshot?.data(), let name = data["name"] as? String, !name.isEmpty {
                    self.currentUserName = name
                    self.needsOnboarding = false
                } else {
                    self.needsOnboarding = true
                }
            }
        }
    }

    // MARK: - Send OTP

    func sendOTP(to phoneNumber: String, completion: @escaping (Result<Void, Error>) -> Void) {
        PhoneAuthProvider.provider().verifyPhoneNumber(phoneNumber, uiDelegate: nil) { [weak self] verificationID, error in
            DispatchQueue.main.async {
                if let error {
                    completion(.failure(error))
                    return
                }
                self?.verificationId = verificationID
                completion(.success(()))
            }
        }
    }

    // MARK: - Verify OTP

    func verifyOTP(_ code: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let verificationId else {
            completion(.failure(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "No verification ID. Please send OTP again."])))
            return
        }

        let credential = PhoneAuthProvider.provider().credential(withVerificationID: verificationId, verificationCode: code)

        Auth.auth().signIn(with: credential) { [weak self] result, error in
            DispatchQueue.main.async {
                if let error {
                    completion(.failure(error))
                    return
                }

                guard let user = result?.user else {
                    completion(.failure(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Sign in failed."])))
                    return
                }

                // Create user document if it doesn't exist
                self?.db.collection("users").document(user.uid).getDocument { snapshot, _ in
                    if snapshot?.exists != true {
                        self?.db.collection("users").document(user.uid).setData([
                            "phone": user.phoneNumber ?? "",
                            "name": "",
                            "groups": [String](),
                            "createdAt": FieldValue.serverTimestamp()
                        ])
                    }
                }

                completion(.success(()))
            }
        }
    }

    // MARK: - Complete Onboarding

    func completeOnboarding(name: String?) {
        let userName = name ?? "Parent"
        self.currentUserName = userName
        self.needsOnboarding = false

        guard !currentUserId.isEmpty else { return }
        db.collection("users").document(currentUserId).updateData([
            "name": userName
        ])
    }

    // MARK: - Sign Out

    func signOut() {
        try? Auth.auth().signOut()
    }

    // MARK: - Delete Account

    func deleteAccount(completion: @escaping (Result<Void, Error>) -> Void) {
        guard let user = Auth.auth().currentUser else {
            completion(.failure(NSError(domain: "", code: -1,
                userInfo: [NSLocalizedDescriptionKey: "No authenticated user found."])))
            return
        }

        let userId = user.uid

        // Step 1: Delete Firestore user document
        db.collection("users").document(userId).delete { error in
            if let error {
                completion(.failure(error))
                return
            }

            // Step 2: Delete Firebase Auth account
            user.delete { error in
                DispatchQueue.main.async {
                    if let error = error as NSError? {
                        if error.code == AuthErrorCode.requiresRecentLogin.rawValue {
                            completion(.failure(NSError(domain: "", code: error.code,
                                userInfo: [NSLocalizedDescriptionKey:
                                "For security, please sign out and sign back in before deleting your account."])))
                        } else {
                            completion(.failure(error))
                        }
                        return
                    }
                    // Auth state listener will reset isAuthenticated = false
                    completion(.success(()))
                }
            }
        }
    }
}

File: android/app/src/main/java/com/classnotes/app/service/AuthService.kt
package com.classnotes.app.service

import android.app.Activity
import com.classnotes.app.AppMode
import com.google.firebase.FirebaseException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.PhoneAuthOptions
import com.google.firebase.auth.PhoneAuthProvider
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FieldValue
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.tasks.await
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

open class AuthService {
    protected val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    protected val _needsOnboarding = MutableStateFlow(false)
    val needsOnboarding: StateFlow<Boolean> = _needsOnboarding

    protected val _currentUserId = MutableStateFlow("")
    val currentUserId: StateFlow<String> = _currentUserId

    protected val _currentUserName = MutableStateFlow("")
    val currentUserName: StateFlow<String> = _currentUserName

    protected val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    protected val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    protected var verificationId: String? = null

    private val auth: FirebaseAuth? by lazy {
        if (AppMode.isDemo) null else FirebaseAuth.getInstance()
    }

    private val db: FirebaseFirestore? by lazy {
        if (AppMode.isDemo) null else FirebaseFirestore.getInstance()
    }

    // Activity reference needed for Phone Auth reCAPTCHA
    var activity: Activity? = null

    init {
        if (!AppMode.isDemo) {
            FirebaseAuth.getInstance().addAuthStateListener { firebaseAuth ->
                val user = firebaseAuth.currentUser
                if (user != null) {
                    _currentUserId.value = user.uid
                    _isAuthenticated.value = true
                    fetchUserProfile(user.uid)
                    NotificationService.updateFCMToken(user.uid)
                } else {
                    _currentUserId.value = ""
                    _currentUserName.value = ""
                    _isAuthenticated.value = false
                }
            }
        }
    }

    private fun fetchUserProfile(userId: String) {
        db?.collection("users")?.document(userId)?.get()
            ?.addOnSuccessListener { snapshot ->
                val data = snapshot.data
                if (data != null) {
                    val name = data["name"] as? String ?: ""
                    if (name.isNotEmpty()) {
                        _currentUserName.value = name
                        _needsOnboarding.value = false
                    } else {
                        _needsOnboarding.value = true
                    }
                } else {
                    _needsOnboarding.value = true
                }
            }
            ?.addOnFailureListener {
                _needsOnboarding.value = true
            }
    }

    open suspend fun sendOTP(phoneNumber: String) {
        val currentAuth = auth ?: throw Exception("Firebase not available in demo mode")
        val currentActivity = activity ?: throw Exception("Activity not set for Phone Auth")

        suspendCoroutine { continuation ->
            val options = PhoneAuthOptions.newBuilder(currentAuth)
                .setPhoneNumber(phoneNumber)
                .setTimeout(60L, TimeUnit.SECONDS)
                .setActivity(currentActivity)
                .setCallbacks(object : PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
                    override fun onVerificationCompleted(credential: PhoneAuthCredential) {
                        // Auto-verification (rare on emulator, common on real devices)
                    }

                    override fun onVerificationFailed(e: FirebaseException) {
                        continuation.resumeWithException(e)
                    }

                    override fun onCodeSent(
                        verificationId: String,
                        token: PhoneAuthProvider.ForceResendingToken
                    ) {
                        this@AuthService.verificationId = verificationId
                        continuation.resume(Unit)
                    }
                })
                .build()
            PhoneAuthProvider.verifyPhoneNumber(options)
        }
    }

    open suspend fun verifyOTP(code: String) {
        val currentAuth = auth ?: throw Exception("Firebase not available in demo mode")
        val vid = verificationId ?: throw Exception("No verification ID. Please send OTP again.")

        val credential = PhoneAuthProvider.getCredential(vid, code)
        val result = currentAuth.signInWithCredential(credential).await()
        val user = result.user ?: throw Exception("Sign in failed.")

        // Create user document if it doesn't exist
        val userDoc = db?.collection("users")?.document(user.uid)?.get()?.await()
        if (userDoc?.exists() != true) {
            db?.collection("users")?.document(user.uid)?.set(
                hashMapOf(
                    "phone" to (user.phoneNumber ?: ""),
                    "name" to "",
                    "groups" to listOf<String>(),
                    "createdAt" to FieldValue.serverTimestamp()
                )
            )?.await()
        }
    }

    open fun completeOnboarding(name: String?) {
        val userName = name?.trim()?.takeIf { it.isNotEmpty() } ?: "Parent"
        _currentUserName.value = userName
        _needsOnboarding.value = false

        val userId = _currentUserId.value
        if (userId.isNotEmpty()) {
            db?.collection("users")?.document(userId)?.update("name", userName)
        }
    }

    open suspend fun updateName(name: String) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) throw Exception("Name cannot be empty.")
        val userId = _currentUserId.value
        if (userId.isEmpty()) throw Exception("Not signed in.")

        db?.collection("users")?.document(userId)?.update("name", trimmed)?.await()
        _currentUserName.value = trimmed
    }

    open fun signOut() {
        auth?.signOut()
        _isAuthenticated.value = false
        _currentUserId.value = ""
        _currentUserName.value = ""
        _needsOnboarding.value = false
    }

    open suspend fun deleteAccount() {
        val user = auth?.currentUser ?: throw Exception("No authenticated user found.")
        val userId = user.uid

        // Step 1: Delete Firestore user document
        db?.collection("users")?.document(userId)?.delete()?.await()

        // Step 2: Delete Firebase Auth account
        try {
            user.delete().await()
        } catch (e: Exception) {
            throw Exception("For security, please sign out and sign back in before deleting your account.")
        }
        // Auth state listener will reset isAuthenticated = false
    }

    fun clearError() {
        _error.value = null
    }
}
